src/Microsoft.DotNet.Build.Tasks.Feed/BlobFeedAction.cs

```diff
@@ -0,0 +1,154 @@
+        private static readonly CancellationToken CancellationToken = TokenSource.Token;
+
+        public BlobFeed feed;
+        public int MaxClients { get; set; } = 8;
+
+        public BlobFeedAction(string accountName, string accountKey, string containerName, string indexDirectory, MSBuild.TaskLoggingHelper Log)
+        {
+            this.feed = new BlobFeed(accountName, accountKey, containerName, string.IsNullOrWhiteSpace(indexDirectory) ? Path.GetTempPath() : indexDirectory, Log);
+            this.Log = Log;
+        }
+
+        public async Task<bool> PushToFeed(IEnumerable<string> items, string relativePath, bool containerExists = false)
+        {
+            if (feed.IsSanityChecked(items))
+            {
+                if (CancellationToken.IsCancellationRequested)
+                {
+                    Log.LogError("Task PushToFeed cancelled");
+                    CancellationToken.ThrowIfCancellationRequested();
+                }
+
+                List<string> indexList = new List<string>(Directory.EnumerateFiles(Path.Combine(feed.IndexDirectory, relativePath), "index.json", SearchOption.AllDirectories));
+
+                using (var clientThrottle = new SemaphoreSlim(this.MaxClients, this.MaxClients))
+                {
+                    await Task.WhenAll(items.Select(item => PushItemToFeed(item, relativePath, clientThrottle)));
+                    if (containerExists)
+                    {
+                        await Task.WhenAll(items.Select(item => UpdateIndexes(item, relativePath)));
+                    }
+                    await Task.WhenAll(indexList.Select(item => PushItemToFeed(item, relativePath, clientThrottle)));
+                }
+            }
+
+            return !Log.HasLoggedErrors;
+        }
+
+        public async Task<bool> PushToFeedFlat(IEnumerable<string> items, string relativePath)
+        {
+            if (CancellationToken.IsCancellationRequested)
+            {
+                Log.LogError("Task PushToFeedFlat cancelled");
+                CancellationToken.ThrowIfCancellationRequested();
+            }
+
+            using (var clientThrottle = new SemaphoreSlim(this.MaxClients, this.MaxClients))
+            {
+                await Task.WhenAll(items.Select(item => PushItemToFeed(item, relativePath, clientThrottle)));
+            }
+            return !Log.HasLoggedErrors;
+        }
+
+        public async Task<bool> PushItemToFeed(string item, string relativePath, SemaphoreSlim clientThrottle)
+        {
+            string uploadPath = feed.CalculateBlobPath(item, relativePath);
+            List<string> listAzureBlobs = await ListAzureBlobs.ListBlobs(Log, feed.AccountName, feed.AccountKey, feed.ContainerName, feed.CalculateRelativeUploadPath(item, relativePath));
+            await UploadAsync(CancellationToken, item, uploadPath, clientThrottle, listAzureBlobs.Contains(uploadPath));
```

Comment:
Move the list after you upload the package and add an assertion that the package you just uploaded is in that list.